File: 2020/ts/day09.ts
import * as fs from "fs";

export function read_input(): Array<number> {
  return fs
    .readFileSync("../inputs/input09", "utf8")
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0)
    .map((line) => parseInt(line));
}

function combinations(nums: Array<number>): Array<[number, number]> {
  let s = new Set<String>();

  for (let i = 0; i < nums.length - 1; i++) {
    for (let j = 1; j < nums.length; j++) {
      if (!s.has(`${nums[j]}:${nums[i]}`)) s.add(`${nums[i]}:${nums[j]}`);
    }
  }

  const res = [];
  s.forEach((v) => {
    const tokens = v.split(":");
    res.push([parseInt(tokens[0]), parseInt(tokens[1])]);
  });
  return res;
}

export function is_valid(nums: Array<number>, target: number) {
  const combs = combinations(nums);
  for (let [i, j] of combs) {
    if (i + j == target) return true;
  }
  return false;
}

export function run(nums: Array<number>, preamble: number = 5): number {
  for (let i = 0; i < nums.length - preamble; i++) {
    const end = i + preamble;
    const n = nums.slice(i, end);
    const target = nums[end];
    if (!is_valid(n, target)) return target;
  }
  return 0;
}

function part1() {
  console.log(run(read_input(), 25));
}

function part2() {
  const nums = read_input();
  const target = 88311122;
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      const sum = nums.slice(i, j).reduce((acc, v) => acc + v, 0);
      if (sum == target) {
        console.log(
          Math.min.apply(Math, nums.slice(i, j)) +
            Math.max.apply(Math, nums.slice(i, j))
        );
        return;
      } else if (sum > target) break;
    }
  }
}

part1();
part2();

File: 2020/python/day09.py
import unittest
import itertools


def valid(nums, target):
    combinations = itertools.combinations(nums, 2)
    return target in (sum(c) for c in combinations)


def run(nums, preamble=5):
    i = 0
    for i in range(0, len(nums) - preamble):
        end = i + preamble
        n = nums[i:end]
        target = nums[end]
        if not valid(n, target):
            return target
    return 0


def part1():
    print(run(read_input(), 25))


def run2(nums, target):
    for i in range(0, len(nums)):
        for j in range(2, len(nums)):
            s = sum(nums[i:j])
            if s == target:
                return min(nums[i:j]) + max(nums[i:j])
            elif s > target:
                break


def part2():
    print(run2(read_input(), 88311122))


def read_input():
    with open('../inputs/input09') as input:
        return [int(l) for l in input if not l.isspace()]


test_input = """35
20
15
25
47
40
62
55
65
95
102
117
150
182
127
219
299
277
309
576"""

test_input2 = """35
20
15
25
47
40
62
55
65
95
102
117
150
182
127
219
299
277
309
576"""


class TestPart1(unittest.TestCase):
    def test_valid(self):
        nums = [num for num in range(1, 26)]
        self.assertTrue(valid(nums, 26))
        self.assertTrue(valid(nums, 49))
        self.assertFalse(valid(nums, 100))
        self.assertFalse(valid(nums, 50))

    def test_run(self):
        nums = [int(l) for l in test_input.split('\n')]
        self.assertEqual(run(nums), 127)


class TestPart2(unittest.TestCase):
    def test_run2(self):
        nums = [int(l) for l in test_input2.split('\n')]
        self.assertEqual(run2(nums, 127), 62)


if __name__ == '__main__':
    part1()
    part2()
    unittest.main()
